### PR TITLE
Initial version of the Quarkus stack [Replacing PR #94]

### PR DESCRIPTION
**Pull request by paulrobinson**
# Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

# Contributing a new stack:

## Describe how application dependencies are managed:
They are managed via the `properties` section of the `pom.xml` in the template code.

## Explain how Appsody file watcher is utilized:
I don't think it is used. I omitted the `APPSODY_WATCH_` variables as Quarkus will watch for change itself and perform live reload. Using the `APPSODY_MOUNTS` variable is enough to mount the developer's source code for Quarkus to watch.

## Describe other Appsody environment variables defined by the stack image:

`ENV APPSODY_RUN="./mvnw -Dmaven.repo.local=/mvn/repository compile quarkus:dev"`
This starts Quarkus in developer mode (live reload and remote debug enabled). It also configures Maven to use the `APPSODY_DEPS` location for the local repo cache. This allows maven to cache dependencies between container runs. 

`ENV APPSODY_DEBUG="./mvnw -Dmaven.repo.local=/mvn/repository -Ddebug compile quarkus:dev"`
This is the same as `APPSODY_RUN` but the Quarkus runtime pauses waiting for the debugger to attach.

`ENV APPSODY_TEST="./mvnw -Dmaven.repo.local=/mvn/repository test"`
This starts Quarkus and runs the tests. However, they currently fail. See 'known limitations' discussed below. 

## Describe any limitations and known issues:
I'm yet to test the `appsody deploy` command, as I don't yet have a Kubernetes cluster that's configured to deploy to.

There's also a problem with `appsody test`. The tests fail due to them not being able to communicate with the Quarkus server. It seems like the `8080` port is not being exposed.

I think both of these issues will need addressing before the PR is merged.

# Related Issues:
None.

### CODE REVIEW

**Comment by mhamwala** [comment]
Thanks @paulrobinson I have looked though, reviewed and tested, looks good so far, the development process works though I’m wondering whether it would be possible for you to move some aspects of the application over to the stack image.

As a stack provider, It would be good if we could add some value to the stack image by providing the capability for the developers to not have to worry about certain aspects such as the Java version or Quarkus version. The value that this capability would add is that is `1.` It simplifies the experience for developers and relieve them of full-stack expertise and responsibilities. `2.` It provides architects the ability to create configurable and opinionated set of technologies. Additionally it provides a place for the developers to pick up new version bumps and provides consistency for each developer using the Quarkus stack.

We could achieve this by including a `image/project/pom.xml` at in the stack image and having the `template/default/pom.xml` in the default template extend it though via a parent pom relationship: for example this has been done in the java-microprofile stack [here](https://github.com/appsody/stacks/tree/master/incubator/java-microprofile). 

Thoughts? @appsody/stacks

**Comment by paulrobinson** [comment]
Thanks @mhamwala. I agree with your feedback, I'll take a look at the example and hopefully make the change this week.

**Comment by paulrobinson** [comment]
@mhamwala Do you know how the child pom.xml (https://github.com/appsody/stacks/blob/master/incubator/java-microprofile/templates/default/pom.xml) finds the parent pom.xml (https://github.com/appsody/stacks/blob/master/incubator/java-microprofile/image/project/pom.xml)?

**Comment by mhamwala** [comment]
@paulrobinson I believe we do this by adding the `<parent>` field and updating the template `pom.xml` to reference the Stack image as shown below.

```
<parent>
        <groupId>dev.appsody</groupId>
        <artifactId>java-microprofile</artifactId>
        <version>0.2.3</version>
</parent>
```

**Comment by mhamwala** [comment]
We also added a way for the stack image to enforce the template application to extend the parent `pom.xml`, feel free to ping me on the Appsody slack and I'll run through this

**Comment by paulrobinson** [comment]
Sorry @mhamwala, I wasn't very clear. I have the parent declared in the child pom.xml. However, I'm unclear on how Maven is finding the parent pom.xml file. In the Microprofile example there's no 'relativePath' element defined in the child pom.xml and the pom.xml doesn't seem to get 'built' and thus available via the maven repo. Also, I couldn't see how the parent pom.xml file is making its way into the container image.

I can chat on Slack if it would be easier.

**Comment by paulrobinson** [comment]
@mhamwala I think I now have the dependencies moved into a parent pom.xml as you suggest. `appsody run` and `appsody build` are working for me with this setup.

Do you think I have moved enough into the parent pom.xml?

**Comment by mhamwala** [comment]
Thanks @paulrobinson Just testing your latest changes now, seems like your tests are failing, it can't hit the `testHelloEndpoint`

**Comment by mhamwala** [comment]
I think as long we provide the capability to control or easily add elements to the stack image, we can always improve on this, I'll just finish reviewing now!

**Comment by paulrobinson** [comment]
I forgot I was having problems with `appsody test` (see the 'known limitations' section in the PR description). 

I'll see if I can get it fixed today or tomorrow. I'm on holiday after that for a week an a half.

**Comment by paulrobinson** [comment]
I've just tried `appsody test` and the tests are passing intermittently. See:

Passing: https://pastebin.com/WG9pn9Yb.
Failing: https://pastebin.com/6T0aGsJX

I'm guessing the tests are not waiting for the server to boot. I'll investigate.

**Comment by mhamwala** [comment]
I'll can test the `appsody deploy` command on my docker for mac kubernetes cluster

**Comment by mhamwala** [comment]
So far reviewed and tested:

- [x] appsody run
- [ ] appsody test
- [ ] appsody debug 
- [x] appsody build
- [x] appsody deploy

**Comment by mhamwala** [comment]
@paulrobinson we could potentially add the following to the parent pom to enforce the java version the child pom uses, using the maven enforcer plugin:
```
<build>
        <plugins>
            <!-- Enforcing JDK Version -->
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-enforcer-plugin</artifactId>
                <version>3.0.0-M2</version>
                <executions>
                <execution>
                    <id>enforce-versions</id>
                    <goals>
                        <goal>enforce</goal>
                    </goals>
                    <configuration>
                        <rules>
                            <requireJavaVersion>
                                <version>[1.8,1.9)</version>
                            </requireJavaVersion>
                        </rules>
                    </configuration>
                </execution>
            </executions>
            </plugin>
        </plugins>
    </build>
```

**Comment by mhamwala** [comment]
@paulrobinson If the `appsody test` or `debug` issue seems to be a bigger issue, then for the meantime we can PR the qaurkus stack into the experimental repo so that we can get some feedback whilst your away.

We can then move it to the incubator repo once those issues are resolved.?

**Comment by paulrobinson** [comment]
Agreed, that would work well for me. I'm running out of time to get this fixed today.

Do I need to make any changes to this PR to move it to the incubator area?

**Comment by mhamwala** [comment]
If you can add the above code block (Java version enforcement), that would be great but atm lgtm if you can create a PR into the `incubator` repo.

File: incubator/quarkus/stack.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+id: quarkus
```
Comment (mhamwala) [inline incubator/quarkus/stack.yaml]: Could you edit this to use the following format:
```
name: Quarkus
version: 0.1.0
description: Quarkus runtime for running Java applications
license: Apache-2.0
maintainers: 
  - name: Paul Robinson 
    email: probinso@redhat.com
    github-id: paulrobinson
default-template: default
```

**Pull request merged**